org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/ClasspathJrt.java

```diff
@@ -149,10 +151,20 @@ public boolean equals(Object o) {
 
 @Override
 public NameEnvironmentAnswer findClass(String binaryFileName, String qualifiedPackageName, String moduleName, String qualifiedBinaryFileName,
-										boolean asBinaryOnly, Predicate<String> moduleNameFilter) {
+										boolean asBinaryOnly, Predicate<String> moduleNameFilter,int releaseNumber) {
 	if (!isPackage(qualifiedPackageName, moduleName)) {
 		return null; // most common case
 	}
+	//TODO TEMP properly merge with ClasspathJrtWithReleaseOption
+	if (releaseNumber>= IReleaseAwareNameEnvironment.FIRST_MULTI_RELEASE) {
+		try {
+			 ClasspathJrtWithReleaseOption withReleaseOption = new ClasspathJrtWithReleaseOption(this.zipFilename, this.accessRuleSet,  this.externalAnnotationPath ==null?null:IPath.fromOSString(this.externalAnnotationPath), Integer.toString(releaseNumber));
```

Comment:
I don't think we can afford to instantiate a `ClasspathJrtWithReleaseOption` for each lookup. To much caching inside that would go to waste.

Reply:
Don't worry this was just a quick one to show it could work, now I need to make it not only work by being well integrated and nicely cached. With this change at least your described case with Java 17 using an API only available in Java 20+ will work (or more formally give an compile error) while it will be compile successful if you use it in a Java 20+ release folder.

Reply:
@stephan-herrmann I now was able to demangle the `ClasspathJrtWithReleaseOption` and the `ClasspathJrt` to support a generic (cached) lookup of release based classe where the first one now is more a lightweight redirect to the super classes.